Make SparseCsr a functionality dispatch key

pearu opened the pull request:
As in the title.

To enable meta and fake tensor support for sparse compressed tensors in compliance with the meta/fake tensor support for sparse COO tensor.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120920
* #120707
* __->__ #120703



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip

pearu committed:
Make SparseCsr a functionality dispatch key

[ghstack-poisoned]

pearu committed:
Update on "Make SparseCsr a functionality dispatch key"




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Make SparseCsr a functionality dispatch key"




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Make SparseCsr a functionality dispatch key"

As in the title.

To enable meta and fake tensor support for sparse compressed tensors in compliance with the meta/fake tensor support for sparse COO tensor.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Make SparseCsr a functionality dispatch key"

As in the title.

To enable meta and fake tensor support for sparse compressed tensors in compliance with the meta/fake tensor support for sparse COO tensor.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -458,6 +462,7 @@ class DispatchKeySet final {
         return 2;
       case DispatchKey::SparseCPU:
         return 3;
+        // TODO: add SparseCsrCPU?
```
No not needed, this is for mobile but I doubt they shipping csr kernels

pearu committed:
Update on "Make SparseCsr a functionality dispatch key"

As in the title.

To enable meta and fake tensor support for sparse compressed tensors in compliance with the meta/fake tensor support for sparse COO tensor.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -2035,8 +2040,21 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+          {BackendComponent::CPUBit,
+           BackendComponent::CUDABit,
+           BackendComponent::HIPBit,
+           BackendComponent::XPUBit});
```
Why do you need to test the backend bits? Can't you just test the functionality key only?

ezyang reviewed (approved):
please check the substantive comment, but looks good!

pearu committed:
Update on "Make SparseCsr a functionality dispatch key"

As in the title.

To enable meta and fake tensor support for sparse compressed tensors in compliance with the meta/fake tensor support for sparse COO tensor.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]
